c10/util/typeid.h

```diff
@@ -442,29 +458,115 @@ class C10_API TypeMeta final {
 #endif
   }
 
- private:
-  const detail::TypeMetaData* data_;
+  /**
+  * convert ScalarType enum values to TypeMeta handles
+  */
+  static inline caffe2::TypeMeta fromScalarType(ScalarType scalar_type) {
+    const size_t index = static_cast<uint16_t>(scalar_type);
+    if (C10_LIKELY(index < NumScalarTypes)) {
+      return TypeMeta(index);
+    }
+    error_unrecognized_scalartype(scalar_type);
+    return TypeMeta(); // avoids "control reaches end of non-void function"
+  }
 
+  /**
+   * convert TypeMeta handles to ScalarType enum values
+   */
+  inline ScalarType toScalarType() {
+    if (C10_LIKELY(isScalarType())) {
+      return static_cast<ScalarType>(index_);
+    }
+    error_unsupported_typemeta(*this);
+    return ScalarType::Undefined; // avoids "control reaches end of non-void function"
+  }
+
+private:
+  static void error_unrecognized_scalartype(ScalarType scalar_type) {
```

Comment:
Use `[[noreturn]]` to mark this as noreturn and then avoid "control reaches end of non-void function" above.